src/psyclone/tests/psyir/transformations/replace_induction_variables_test.py

```diff
@@ -0,0 +1,188 @@
+def test_riv_working(fortran_reader, fortran_writer):
+    '''Tests if loop-invariant assignments are replaced as expected.'''
+    source = '''program test
+                integer i, invariant, ic1, ic2, ic3, ic4, ic5, ic6
+                real, dimension(10) :: a
+                invariant = 1
+                do i = 1, 10
+                    ic1 = 12                ! Constant
+                    ic2 = invariant         ! Invariant variable
+                    ic3 = i+1               ! Dependent on loop variable
+                    ic4 = i + 1 + invariant ! Loop variable plus invariant
+                    ic5 = i*i+3*sin(i)      ! Complicated expression
+                    ic6 = ic4 + ic5         ! Multi-step replacement
+                    a(ic1) = 1+(ic1+1)*ic1
+                    a(ic2) = 2+(ic2+1)*ic2
+                    a(ic3) = 3+(ic3+1)*ic3
+                    a(ic4) = 4+(ic4+1)*ic4
+                    a(ic5) = 5+(ic5+1)*ic5
+                    a(ic6) = 6+(ic6+1)*ic6
+                end do
+                end program test'''
+    psyir = fortran_reader.psyir_from_source(source)
+    # The first child is the assignment to 'invariant'
+    loop = psyir.children[0].children[1]
+
+    riv = ReplaceInductionVariables()
+    riv.apply(loop)
+    out_all = fortran_writer(psyir)
+    out_loop = fortran_writer(loop)
+
+    assert "a(12) = 1 + (12 + 1) * 12" in out_loop
+    # Make sure the assignment to ic1 has been added outside of the loop:
+    assert "ic1 = 12" in out_all
+    assert "ic1 = 12" not in out_loop
+
+    assert "a(invariant) = 2 + (invariant + 1) * invariant" in out_loop
+    # Make sure the assignment to ic2 has been added outside of the loop:
+    assert "ic2 = invariant" in out_all
+    assert "ic2 = invariant" not in out_loop
+
+    assert "a(i + 1) = 3 + (i + 1 + 1) * (i + 1)" in out_loop
+    assert "ic3 = i - 1 + 1" in out_all
+
+    assert ("a(i + 1 + invariant) = 4 + (i + 1 + invariant + 1) * "
+            "(i + 1 + invariant)" in out_loop)
+    assert "ic4 = i - 1 + 1 + invariant" in out_all
+
+    assert ("a(i * i + 3 * SIN(i)) = 5 + (i * i + 3 * SIN(i) + 1) * "
+            "(i * i + 3 * SIN(i))" in out_loop)
+    assert "ic5 = (i - 1) * (i - 1) + 3 * SIN(i - 1)" in out_all
+    assert ("a(i + 1 + invariant + (i * i + 3 * SIN(i))) = 6 + "
+            "(i + 1 + invariant + (i * i + 3 * SIN(i)) + 1) * "
+            "(i + 1 + invariant + (i * i + 3 * SIN(i)))" in out_loop)
+    assert ("ic6 = i - 1 + 1 + invariant + ((i - 1) * (i - 1) + "
+            "3 * SIN(i - 1))" in out_all)
+
+
+# ----------------------------------------------------------------------------
+def test_riv_not_movable(fortran_reader, fortran_writer):
+    '''Tests assignments that cannot be moved.
+    '''
+    source = '''program test
+                integer i, ic1, ic2, ic3, ic4, ic5
+                real, dimension(10) :: a
+                do i = 1, 10
```

Comment:
can we also add an induction variable with a function call in the RHS. This should not be transformed because we don't know if it has side effects that are sensitive to the number of times they are called. (Note this is not equivalent to the sin() in the previous test as sin should be a Operation and it is fine because all intrinsic operations that we support are pure functions)

Reply:
Done. As mentioned previously, I had to modify the PSyIR, I couldn't get it to read in a function (it ended up a codeblock only, not a call).